bgpd/bgp_lcommunity.c

```diff
@@ -396,12 +398,20 @@ static const char *lcommunity_gettoken(const char *str,
 				}
 			} else {
 				digit = 1;
-				localdata2 *= 10;
-				localdata2 += (*p - '0');
+				/* multiply by 10 and add the next digit */
+				increment = 9 * localdata2 + (*p - '0');
+				if (((uint32_t)(localdata2 + increment))
```

Comment:
nit: this will work if we check against UINT_MAX:

```
if (localdata2 > UINT_MAX) {
  *token = lcommunity_token_unknown;
  return NULL;
}
```
Can you test it?

Reply:
This won't work, as the sum accumulates in a `uint32_t` and is consequently always less than `UINT_MAX`.
```
(gdb) p localdata2
$3 = 4294967294
(gdb) p increment
$4 = 4
(gdb) p UINT_MAX
$5 = 4294967295
(gdb) p localdata2 + increment > UINT_MAX
$6 = 0
```
... actually the original explicit typing is already implied so this can be reduced without loss of information to `if (localdata2 + increment < localdata2)`. This should not cause any weirdness on any platforms as no rvalues are cast to anything other than their native type.

Reply:
I was about this:

```
diff --git a/bgpd/bgp_lcommunity.c b/bgpd/bgp_lcommunity.c
index 5294d6e39..66c2c4faf 100644
--- a/bgpd/bgp_lcommunity.c
+++ b/bgpd/bgp_lcommunity.c
@@ -375,7 +375,6 @@ static const char *lcommunity_gettoken(const char *str,
        if (isdigit((unsigned char)*p)) {
                int separator = 0;
                int digit = 0;
-               uint32_t increment = 0;
                uint32_t globaladmin = 0;
                uint32_t localdata1 = 0;
                uint32_t localdata2 = 0;
@@ -398,15 +397,14 @@ static const char *lcommunity_gettoken(const char *str,
                                }
                        } else {
                                digit = 1;
+                               localdata2 *= 10;
+                               localdata2 += (*p - '0');
                                /* multiply by 10 and add the next digit */
-                               increment = 9 * localdata2 + (*p - '0');
-                               if (((uint32_t)(localdata2 + increment))
-                                   < localdata2) {
+                               if (localdata2 > UINT_MAX) {
                                        /* unsigned overflow */
                                        *token = lcommunity_token_unknown;
                                        return NULL;
                                }
-                               localdata2 += increment;
                        }
                        p++;
                }
```
Testing:
```
root@exit1-debian-9:~/frr# vtysh -c 'c' -c 'bgp large-community-list standard test permit 4294967295:4294967296:0'
% Malformed community-list value
root@exit1-debian-9:~/frr# vtysh -c 'c' -c 'bgp large-community-list standard test permit 4294967295:4294967295:0'
root@exit1-debian-9:~/frr# vtysh -c 'c' -c 'bgp large-community-list standard test permit 4294967295:4294967295'
% Malformed community-list value
root@exit1-debian-9:~/frr# vtysh -c 'c' -c 'bgp large-community-list standard test permit 4294967295:4294967295:4294967296'
% Malformed community-list value
root@exit1-debian-9:~/frr# vtysh -c 'c' -c 'bgp large-community-list standard test permit 4294967295:4294967295:123'
```

Reply:
Oh got it, I like that more

Reply:
Updated to use these semantics

Reply:
This does not fix the issue actually, as the vtysh command you're testing with only tests large community lists which work fine on master. This patch fixes the `set large-community` under a `route-map` directive.

```
# Without this patch
wcoakley@fss:~$ vtysh -c 'c' -c 'route-map test permit 1' -c 'set large-community 4294967296:0:0'
wcoakley@fss:~$ echo $?
0
# git checkout, rebuild etc. etc.
# ...with this patch
wcoakley@fss:~$ vtysh -c 'c' -c 'route-map test permit 1' -c 'set large-community 4294967296:0:0'
% [BGP] Argument form is unsupported or malformed.
```

Reverting this patch to the original code since that works as described in the PR

Reply:
Ah, OK, then we need to fix this inside a caller function. We have already validation function which should be called before passing to tokenizing, please test this patch:

```
diff --git a/bgpd/bgp_clist.c b/bgpd/bgp_clist.c
index 247d758f8..9f6741bf7 100644
--- a/bgpd/bgp_clist.c
+++ b/bgpd/bgp_clist.c
@@ -1110,7 +1110,7 @@ struct lcommunity *lcommunity_list_match_delete(struct lcommunity *lcom,
 }
 
 /* Helper to check if every octet do not exceed UINT_MAX */
-static bool lcommunity_list_valid(const char *community)
+bool lcommunity_list_valid(const char *community)
 {
        int octets;
        char **splits, **communities;
diff --git a/bgpd/bgp_clist.h b/bgpd/bgp_clist.h
index f7d46525a..5502a5709 100644
--- a/bgpd/bgp_clist.h
+++ b/bgpd/bgp_clist.h
@@ -157,6 +157,7 @@ extern int lcommunity_list_set(struct community_list_handler *ch,
 extern int lcommunity_list_unset(struct community_list_handler *ch,
                                 const char *name, const char *str,
                                 const char *seq, int direct, int style);
+extern bool lcommunity_list_valid(const char *community);
 
 extern struct community_list_master *
 community_list_master_lookup(struct community_list_handler *, int);
diff --git a/bgpd/bgp_lcommunity.c b/bgpd/bgp_lcommunity.c
index 5900fcf86..a7b4b485a 100644
--- a/bgpd/bgp_lcommunity.c
+++ b/bgpd/bgp_lcommunity.c
@@ -442,6 +442,9 @@ struct lcommunity *lcommunity_str2com(const char *str)
        enum lcommunity_token token = lcommunity_token_unknown;
        struct lcommunity_val lval;
 
+       if (!lcommunity_list_valid(str))
+               return NULL;
+
        do {
                str = lcommunity_gettoken(str, &lval, &token);
                switch (token) {
```

Reply:
That's a good idea... I didn't notice we already had a function which validates the exact thing we're looking to validate, haha...

I applied this patch and reworked `lcommunity_gettoken` to expect a list of valid large communities. I also did some work to improve `lcommunity_list_valid` to reject invalid names like `1:2a:3` because `strtoul` will still partially parse that middle value as a 2 (I just check that the char **endptr value is not `\0` for this).

```
wcoakley@fss:~/code/frr$ vtysh -c 'c' -c 'route-map test permit 1' -c 'set large-community 1:2:3'
wcoakley@fss:~/code/frr$ vtysh -c 'c' -c 'route-map test permit 1' -c 'set large-community :2:3'
% [BGP] Argument form is unsupported or malformed.
wcoakley@fss:~/code/frr$ vtysh -c 'c' -c 'route-map test permit 1' -c 'set large-community 1:2a:3'
% [BGP] Argument form is unsupported or malformed.
wcoakley@fss:~/code/frr$ vtysh -c 'c' -c 'route-map test permit 1' -c 'set large-community 1:4294967296:3'
% [BGP] Argument form is unsupported or malformed.
wcoakley@fss:~/code/frr$ vtysh -c 'c' -c 'route-map test permit 1' -c 'set large-community 1:4294967295:3'
wcoakley@fss:~/code/frr$
```